### ISSUE TITLE
RAC 3 LocationManager

### ISSUE THREAD

**Issue by samritchie** [ReactiveCocoa/ReactiveCocoa#1997]
Is there an example anywhere of an RAC 3 LocationManager? Just starting with RAC and I’m struggling to piece together usage/guidance from issues & blog posts - some good example code would help.

**Comment by samritchie** [comment]
This is what I have currently; would appreciate any feedback.

    class LocationDelegate : NSObject, CLLocationManagerDelegate, Disposable {
        private(set) var disposed: Bool = false
        var locationsUpdated : (([CLLocation]) -> ())?
        var locationError: ((NSError) -> ())?
        
        func locationManager(manager: CLLocationManager!, didUpdateLocations locations: [AnyObject]!) {
            if let locs = locations as? [CLLocation] { locationsUpdated?(locs) }
        }
        
        private func locationManager(manager: CLLocationManager!, didFailWithError error: NSError!) {
            locationError?(error)
        }
        
        func dispose() {
            disposed = true
        }
    }

    func currentLocationSignal(accuracy: CLLocationAccuracy) -> SignalProducer<CLLocation, NSError> {
        return SignalProducer<CLLocation, NSError>() { sink, disposable in
            var locationManager = CLLocationManager()
            var locationDelegate = LocationDelegate()
            locationManager.desiredAccuracy = accuracy
            locationManager.delegate = locationDelegate
            locationDelegate.locationsUpdated = { locations in
                if locations.count > 0 {
                    sendNext(sink, locations.last!)
                }
            }
            locationDelegate.locationError = { error in
                sendError(sink, error)
            }
            locationManager.startUpdatingLocation()
            
            disposable.addDisposable(locationDelegate)
            disposable.addDisposable {
                sendCompleted(sink)
                locationManager.stopUpdatingLocation()
            }
        }
    }

**Comment by neilpa** [comment]
I haven't tried it but @ohwutup mentioned his project [wrapping LocationManager](https://github.com/ohwutup/ProximityManager/blob/master/ProximityManager/Managers/LocationManager.swift) the other day.

**Comment by samritchie** [comment]
Great, thanks. Looks like quite a different approach...

**Comment by samritchie** [comment]
If I’m understanding the code correctly, it’s doing a hybrid hot/cold approach, i.e. the `start()` method creates private `SignalProducer`s that publish to public `Signal`s. What would be the rationale behind this design? Just trying to understand what the tradeoffs are & what problems are being solved.

**Comment by ohwutup** [comment]
I only have a minute now, but can try to go into a bit more depth later on today- really, I was just struggling pretty hard to try to get `Signal`s out of the location manager delegate. 

You are right that its sort of a hybrid approach, but the `SignalProducer` from the delegate calls is only there to start sending events to the respective observer (`SinkOf<T>`). Then, any update sent to said observer will be passed through the `Signal` which was created at the same time as the observer via `Signal.pipe()`

Definitely not sure if this is the best approach, but its been a nice way to get hot `Signals` out of the cold `SignalProducers`.

**Comment by samritchie** [comment]
Thanks ohwutup. What’s the benefit of getting hot signals - does it simplify composition further down the line?

**Comment by ohwutup** [comment]
So, I don't think that it necessarily helps with composition down the line, my main reason for doing it this way was both (and mostly) that it worked, and that the way I reasoned about this in my head, getting location updates seem to be a consistent stream of updating information and any subscriber that comes along should only see the current state and updates moving forward rather than creating its own stream of events as I have understood `SignalProducer`s to do.

tl;dr- `CLLocationManager` updates seem to be better suited to hot `Signal`s vs. cold `SignalProducer`s

**Comment by samritchie** [comment]
Ah, okay. I’ve been considering `CLLocationManager` updates as cold given that observation has side-effects (i.e. spins the GPS up & down). I guess it would depend on your app. I’ll have to do some testing & see how iOS handles multiple location managers in practice.

**Comment by neilpa** [comment]
I've written a couple of "warm" `SignalProducer`s which may be appropriate in this scenario. That is, a long running `SignalProducer` that _may_ defer work until the first subscription. Internally it uses a `Signal.pipe()` to manage observers and feed events too. Depending on your situation you may also want to track the "current" state and provide that as an initial value for each new subscription.

A contrived example that hopefully illustrates the idea.
```swift
public class Counter {
    private var count = 0
    private let sink: Signal<Int, NoError>.Observer
    private let signal: Signal<Int, NoError>

    public private(set) lazy var producer: SignalProducer<Int, NoError> = SignalProducer { observer, disposable in
        // Depending on your usage you may need to lock while updating.
        sendNext(observer, self.count)
        self.signal.observe(observer)
    }

    public init() {
        (signal, sink) = Signal<Int, NoError>.pipe()
    }

    public func increment() {
        // Depending on your usage you may need to lock while updating.
        sendNext(sink, ++count)
    }
}
```

**Comment by jspahrsummers** [comment]
cc @robb

**Comment by RuiAAPeres** [comment]
Closing due to inactivity.
